File: alc/lexer.c
// Lexical tokeniser for the alic compiler.
// (c) 2019, 2025 Warren Toomey, GPL3

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include "alic.h"
#include "proto.h"

#define TEXTLEN 512

static int Linestart= 0;
static int Putback= 0;
char Text[TEXTLEN + 1];		// Text of the last token scanned
Token Peektoken;		// A look-ahead token
Token Thistoken;		// The last token scanned

// Get the next character from the input file.
static int next(void) {
  int c, l;

  if (Putback) {		// Use the character put
    c = Putback;		// back if there is one
    Putback = 0;
    return(c);
  }

  c = fgetc(Infh);		// Read from input file

  while (Linestart && c == '#') {	// We've hit a pre-processor statement
    Linestart = 0;			// No longer at the start of the line
    scan(&Thistoken);			// Get the line number into l
    if (Thistoken.token != T_NUMLIT)
      fatal("Expecting pre-processor line number, got %s\n", Text);
    l = Thistoken.numval.intval;

    scan(&Thistoken);			// Get the filename in Text
    if (Thistoken.token != T_STRLIT)
      fatal("Expecting pre-processor file name, got %s\n", Text);

    if (Text[0] != '<') {		// If this is a real filename
      if (strcmp(Text, Infilename))	// and not the one we have now
	Infilename = strdup(Text);	// save it. Then update the line num
      Line = l;
    }

    while ((c = fgetc(Infh)) != '\n'); // Skip to the end of the line
    c = fgetc(Infh);			// and get the next character
    Linestart = 1;			// Now back at the start of the line
  }

  Linestart = 0;			// No longer at the start of the line
  if ('\n' == c) {
    Line++;				// Increment line count
    Linestart = 1;			// Now back at the start of the line
  }
  return(c);
}

// Put back an unwanted character
static void putback(int c) {
  Putback = c;
}

// Skip past input that we don't need to deal with, 
// i.e. whitespace, newlines. Return the first
// character we do need to deal with.
static int skip(void) {
  int c;

  c = next();
  while (isspace(c))
    c = next();
  return(c);
}

// Return the position of character c
// in string s, or -1 if c not found
static int chrpos(char *s, int c) {
  int i;
  for (i = 0; s[i] != '\0'; i++)
    if (s[i] == (char) c)
      return (i);
  return (-1);
}

// Read in a hexadecimal constant from the input
static int hexchar(void) {
  int c, h, n = 0, f = 0;

  // Loop getting characters
  while (isxdigit(c = next())) {
    // Convert from char to int value
    h = chrpos("0123456789abcdef", tolower(c));

    // Add to running hex value
    n = n * 16 + h;
    f = 1;
  }

  // We hit a non-hex character, put it back
  putback(c);

  // Flag tells us we never saw any hex characters
  if (!f)
    fatal("missing digits after '\\x'\n");
  if (n > 255)
    fatal("value out of range after '\\x'\n");

  return(n);
}

// Return the next character from a character
// or string literal
static int scanch(void) {
  int i, c, c2;

  // Get the next input character and interpret
  // metacharacters that start with a backslash
  c = next();
  if (c == '\\') {
    switch (c = next()) {
    case 'a':
      return('\a');
    case 'b':
      return('\b');
    case 'f':
      return('\f');
    case 'n':
      return('\n');
    case 'r':
      return('\r');
    case 't':
      return('\t');
    case 'v':
      return('\v');
    case '\\':
      return('\\');
    case '"':
      return('"');
    case '\'':
      return('\'');

      // Deal with octal constants by reading in
      // characters until we hit a non-octal digit.
      // Build up the octal value in c2 and count
      // # digits in i. Permit only 3 octal digits.
    case '0':
    case '1':
    case '2':
    case '3':
    case '4':
    case '5':
    case '6':
    case '7':
      for (i = c2 = 0; isdigit(c) && c < '8'; c = next()) {
	if (++i > 3)
	  break;
	c2 = c2 * 8 + (c - '0');
      }

      putback(c);		// Put back the first non-octal char
      return(c2);
    case 'x':
      return(hexchar());
    default:
      fatal("unknown escape sequence %c\n", c);
    }
  }
  return(c);			// Just an ordinary old character!
}

// List of characters that can be found in a numeric literal
char *numchar= "0123456789ABCDEFabcdef.x";

// Scan a numeric literal value from the input file
// and store it in the given token pointer
static void scan_numlit(Token *t, int c, bool is_negative) {
  int i=0;
  bool isfloat=false;
  int radix=10;

  // Assume an unsigned int
  t->numtype= NUM_UINT;

  // Put the first character and negative sign in the buffer
  if (is_negative) {
    Text[i++]= '-';
    t->numtype= NUM_INT;
  }
  Text[i++]= c;

  // Loop while we have enough buffer space
  for (; i < TEXTLEN - 1; i++) {
    c = scanch();

    // Found a non-numeric character
    if (strchr(numchar,c)==NULL) {
      putback(c);
      break;
    }

    // Otherwise add it to the buffer
    Text[i]= c;
  }

  // NUL terminate the string
  Text[i] = '\0';

  // Determine either if it's a float
  // or any octal/hex radix
  if (strchr(Text, '.') != NULL) {
    isfloat= true;
    t->numtype = NUM_FLT;
  } else {
    if (Text[0]=='0') {
      if (Text[1]=='x')
	radix=16;
      else
	radix=8;
    }
  }

  // Do the conversion
  if (isfloat)
    t->numval.dblval = strtod(Text, NULL);
  else
    t->numval.uintval = strtoull(Text, NULL, radix);
}

// Scan in a string literal from the input file,
// and store it in buf[]. Return the length of
// the string. 
static int scanstr(char *buf) {
  int i, c;

  // Loop while we have enough buffer space
  for (i = 0; i < TEXTLEN - 1; i++) {
    // Get the next char and append to buf
    // Return when we hit the ending double quote
    if ((c = scanch()) == '"') {
      buf[i] = 0;
      return(i);
    }
    buf[i] = (char) c;
  }

  // Ran out of buf[] space
  fatal("String literal too long\n");
  return(0);
}

// Scan an identifier from the input file and
// store it in buf[]. Return the identifier's length
static int scanident(int c, char *buf, int lim) {
  int i = 0;

  // Allow digits, alpha and underscores
  while (isalpha(c) || isdigit(c) || '_' == c) {
    // Error if we hit the identifier length limit,
    // else append to buf[] and get next character
    if (lim - 1 == i) {
      fatal("Identifier too long\n");
    } else if (i < lim - 1) {
      buf[i++] = (char) c;
    }
    c = next();
  }

  // We hit a non-valid character, put it back.
  // NUL-terminate the buf[] and return the length
  putback(c);
  buf[i] = '\0';
  return(i);
}


// A structure to hold a keyword, its first letter
// and the token id associated with the keyword
struct keynode {
  char first;
  char *keyword;
  int  token;
};

// List of keywords and matching tokens
static struct keynode keylist[]= {
  { 'b', "bool", T_BOOL },
  { 'e', "else", T_ELSE },
  { 'f', "false", T_FALSE },
  { 'f', "flt32", T_FLT32 },
  { 'f', "flt64", T_FLT64 },
  { 'f', "for", T_FOR },
  { 'i', "if", T_IF },
  { 'i', "int8", T_INT8 },
  { 'i', "int16", T_INT16 },
  { 'i', "int32", T_INT32 },
  { 'i', "int64", T_INT64 },
  { 'p', "printf", T_PRINTF },
  { 't', "true", T_TRUE },
  { 'u', "uint8", T_UINT8 },
  { 'u', "uint16", T_UINT16 },
  { 'u', "uint32", T_UINT32 },
  { 'u', "uint64", T_UINT64 },
  { 'v', "void", T_VOID },
  { 'w', "while", T_WHILE },
  { 0,   NULL,    0 }
};

// Given a word from the input, return the matching
// keyword token number or 0 if it's not a keyword.
// Switch on the first letter so that we don't have
// to waste time strcmp()ing against all the keywords.
static int keyword(char *s) {
  int i;

  for (i=0; keylist[i].first != 0; i++) {
    // Too early
    if (keylist[i].first < *s) continue;

    // A match
    if (!strcmp(s, keylist[i].keyword)) return(keylist[i].token);

    // Too late
    if (keylist[i].first > *s) return(0);
  }

  return(0);
}

// Scan and return the next token found in the input.
// Return 1 if token valid, 0 if no tokens left.
int scan(Token * t) {
  int c, tokentype;

  // If we have a lookahead token, return this token
  if (Peektoken.token != 0) {
    t->token = Peektoken.token;
    t->tokstr = Peektoken.tokstr;
    t->numval.intval = Peektoken.numval.intval;
    t->numtype = Peektoken.numtype;
    Peektoken.token = 0;
    return(1);
  }

  // Skip whitespace
  c = skip();

  // Determine the token based on
  // the input character
  switch (c) {
  case EOF:
    t->token = T_EOF;
    return(0);
  case '+':
    t->token = T_PLUS;
    break;
  case '-':
    c= next();

    if (isdigit(c)) {		// Negative numeric literal
      scan_numlit(t, c, true);
      t->token = T_NUMLIT;
    } else {
      putback(c);
      t->token = T_MINUS;
    }
    break;
  case '*':
    t->token = T_STAR;
    break;
  case '/':
    t->token = T_SLASH;
    break;
  case ';':
    t->token = T_SEMI;
    break;
  case '{':
    t->token = T_LBRACE;
    break;
  case '}':
    t->token = T_RBRACE;
    break;
  case '(':
    t->token = T_LPAREN;
    break;
  case ')':
    t->token = T_RPAREN;
    break;
  case '~':
    t->token = T_INVERT;
    break;
  case '^':
    t->token = T_XOR;
    break;
  case ',':
    t->token = T_COMMA;
    break;
  case '=':
    if ((c = next()) == '=') {
      t->token = T_EQ;
    } else {
      putback(c);
      t->token = T_ASSIGN;
    }
    break;
  case '!':
    if ((c = next()) == '=') {
      t->token = T_NE;
    } else {
      putback(c);
      t->token = T_LOGNOT;
    }
    break;
  case '<':
    if ((c = next()) == '=') {
      t->token = T_LE;
    } else if (c == '<') {
      t->token = T_LSHIFT;
    } else {
      putback(c);
      t->token = T_LT;
    }
    break;
  case '>':
    if ((c = next()) == '=') {
      t->token = T_GE;
    } else if (c == '>') {
      t->token = T_RSHIFT;
    } else {
      putback(c);
      t->token = T_GT;
    }
    break;
  case '&':
    if ((c = next()) == '&') {
      t->token = T_LOGAND;
    } else {
      putback(c);
      t->token = T_AMPER;
    }
    break;
  case '|':
    if ((c = next()) == '|') {
      t->token = T_LOGOR;
    } else {
      putback(c);
      t->token = T_OR;
    }
    break;
  case '\'':
    // If it's a quote, scan in the
    // literal character value and
    // the trailing quote
    t->numval.intval = scanch();
    t->numtype= NUM_CHAR;
    t->token = T_NUMLIT;
    if (next() != '\'')
      fatal("Expected '\\'' at end of char literal\n");
    break;
  case '"':
    // Scan in a literal string
    scanstr(Text);
    t->token = T_STRLIT;
    t->tokstr= strdup(Text);
    break;
  default:
    // If it's a digit, scan the
    // literal integer value in
    if (isdigit(c)) {
      scan_numlit(t,c,false);
      t->token = T_NUMLIT;
      break;
    } else if (isalpha(c) || '_' == c) {
      // Read in a keyword or identifier
      scanident(c, Text, TEXTLEN);

      // If it's a recognised keyword, return that token
      if ((tokentype = keyword(Text)) != 0) {
	t->token = tokentype;
	break;
      }
      // Not a recognised keyword, so it must be an identifier
      t->token = T_IDENT;
      t->tokstr= strdup(Text);
      break;
    }

    // The character isn't part of any recognised token, error
    fatal("Unrecognised character: %c\n", c);
  }

  // We found a token
  return(1);
}

// List of tokens as strings
static char *tokstr[]= {
  "EOF",

  "&", "|", "^",
  "==", "!=", "<", ">", "<=", ">=",
  "<<", ">>",
  "+", "-", "*", "/", "%",

  "=", "~", "!", "&&", "||",

  "void",  "bool",
  "int8",  "int16",  "int32",  "int64",
  "uint8", "uint16", "uint32", "uint64",
  "flt32", "flt64",

  "if", "else", "false", "for", "printf",
  "true", "while",

  "numlit", "strlit", ";", "ident",
  "{", "}", "(", ")",
  ","
};

char *get_tokenstr(int token) {
  return(tokstr[token]);
}

// Dump the tokens in the input file
void dumptokens(void) {
  Token t;

  while (1) {
    if (scan(&t)==0) return;
    fprintf(Debugfh, "%s", tokstr[t.token]);
    switch(t.token) {
    case T_STRLIT:
      fprintf(Debugfh, " \"%s\"", Text);
      break;
    case T_NUMLIT:
      if (t.numtype==NUM_CHAR) {
        fprintf(Debugfh, " '%c'", (int)t.numval.intval);
	break;
      }
      // fallthrough
    case T_IDENT:
      fprintf(Debugfh, " %s", Text);
    }
    fprintf(Debugfh, "\n");
  }
}

// Ensure that the current token is t,
// and psossibly fetch the next token.
// Otherwise throw an error
void match(int t, bool getnext) {
  if (Thistoken.token != t)
    fatal("Expected %s, got %s\n", tokstr[t], tokstr[Thistoken.token]);

  if (getnext) scan(&Thistoken);
}

// Match a semicolon and fetch the next token
void semi(void) {
  match(T_SEMI,true);
}

// Match a left brace and fetch the next token
void lbrace(void) {
  match(T_LBRACE,true);
}

// Match a right brace and fetch the next token
void rbrace(void) {
  match(T_RBRACE,true);
}

// Match a left parenthesis and fetch the next token
void lparen(void) {
  match(T_LPAREN,true);
}

// Match a right parenthesis and fetch the next token
void rparen(void) {
  match(T_RPAREN,true);
}

// Match an identifer and fetch the next token
void ident(void) {
  match(T_IDENT,true);
}

// Match a comma and fetch the next token
void comma(void) {
  match(T_COMMA,true);
}

File: alpy/lexer.py
import sys
from defs import TokenType, NumType, Token, fatal, read_file, slash_char

TextLen = 512
Keywords = {
    'void': TokenType.T_VOID,
    'bool': TokenType.T_BOOL,
    'int8': TokenType.T_INT8,
    'int16': TokenType.T_INT16,
    'int32': TokenType.T_INT32,
    'int64': TokenType.T_INT64,
    'uint8': TokenType.T_UINT8,
    'uint16': TokenType.T_UINT16,
    'uint32': TokenType.T_UINT32,
    'uint64': TokenType.T_UINT64,
    'flt32': TokenType.T_FLT32,
    'flt64': TokenType.T_FLT64,
    'if': TokenType.T_IF,
    'else': TokenType.T_ELSE,
    'false': TokenType.T_FALSE,
    'for': TokenType.T_FOR,
    'printf': TokenType.T_PRINTF,
    'true': TokenType.T_TRUE,
    'while': TokenType.T_WHILE
}


class Lexer:
    line_no, at_begin = 1, False
    # 单行注释%%和多行注释<%%>
    in_comment, mul_comment = False, False

    def __init__(self, filename: str):
        self.filename = filename
        self.source = read_file(filename)
        self.reset()

    def reset(self):
        self.pos, self.size = -1, len(self.source)

    @staticmethod
    def ord_char(s: str, c: str) -> int:
        for i, char in enumerate(s):
            if char == c:
                return i
        return -1

    def back_pos(self, n: int = 1):
        self.pos -= n

    def hex_char(self) -> str:
        c = self.next_char().lower()
        regular, n = False, 0
        while c != '\0' and (c.isdigit() or 'a' <= c <= 'f'):
            h = self.ord_char("0123456789abcdef", c)
            regular, n = True, n * 16 + h
            c = self.next_char().lower()
        self.back_pos()
        if not regular:
            fatal("missing digits after '\\x'")
        if n > 255:
            fatal("value out of range after '\\x'")
        return chr(n)

    def unquote_char(self, c: str) -> str:
        if c != '\\':
            return c
        c = self.next_char()
        c2 = slash_char(c)
        if c2 != '':
            return c2

        if '0' <= c <= '7':
            i, c3 = 0, 0
            while c != '\0' and '0' <= c <= '7' and i < 3:
                c3 = c3 * 8 + (ord(c) - ord('0'))
                i += 1
                c = self.next_char()
            self.back_pos()
            return chr(c3)
        if c == 'x':
            return self.hex_char()
        fatal(f"unknown escape sequence {c}")
        return ''

    def next_char(self, ignore_space=False) -> str:
        self.pos += 1
        if self.pos >= self.size:
            return '\0'
        c = self.source[self.pos]
        if c == '\n':
            self.line_no += 1
            self.at_begin = True
            self.in_comment = False
            return self.next_char(True)
        while self.in_comment or ignore_space and c.isspace():
            c = self.next_char(True)
        return c

    def dump_tokens(self, out=None):
        if not out:
            out = sys.stdout
        line = 0
        self.reset()
        out.write("\nTokens in {}".format(self.filename))
        for token in self.scan_all():
            if self.line_no > line:
                line = self.line_no
                out.write("\n{}: ".format(line))
            out.write("{} ".format(token))
        out.write("\n")

    def scan_next(self):
        c = self.next_char(True)
        if c == '\0':  # EOF
            return Token(TokenType.T_EOF)
        while self.mul_comment:
            c = self.next_char(True)
            if self.scan_comment(c):
                c = self.next_char(True)
        # 寻找Token
        if c.isalpha() or c == '_':
            return self.scan_keyword(c)
        elif c == '"' or c == '\'' or c == '\\':
            return self.scan_string(c)
        elif c.isdigit() or c == '.':
            return self.scan_number(c)
        elif not self.scan_comment(c):
            return self.scan_symbol(c)
        return self.scan_next()

    def scan_all(self):
        token = self.scan_next()
        while token.token != TokenType.T_EOF:
            yield token
            token = self.scan_next()
        yield token

    def scan_comment(self, c: str) -> bool:
        c2 = ''
        if self.mul_comment and c == '%':
            c2 = self.next_char(False)
            if c2 == '>':
                self.mul_comment = False
                return True
        if not self.mul_comment and c == '<':
            c2 = self.next_char(False)
            if c2 == '%':
                self.mul_comment = True
                return True
        if not self.mul_comment and c == '%':
            c2 = self.next_char(False)
            if c2 == '%':
                self.in_comment = True
                return True
        if c2 != '':
            self.back_pos()
        return False

    def scan_keyword(self, c: str) -> Token:
        i, buf = 0, []
        while i < TextLen - 1 and (c.isalnum() or c == '_'):
            buf.append(c)
            i += 1
            c = self.next_char()
        self.back_pos()
        word = ''.join(buf)
        tok_type = Keywords.get(word, TokenType.T_IDENT)
        token = Token(tok_type)
        token.tok_str = word
        return token

    def scan_string(self, delim: str) -> Token:
        i, buf = 0, [delim, ]
        c = self.next_char()
        while c != delim and c != '':
            if i >= TextLen - 1:
                fatal("String too long")
            if c == '\\':
                c = self.unquote_char(c)
            buf.append(c)
            i += 1
            c = self.next_char()
        if c == '':
            fatal("Unclosed string")
        else:
            buf.append(delim)
        token = Token(TokenType.T_STRLIT)
        token.tok_str = ''.join(buf)
        return token

    def scan_number(self, c: str) -> Token:
        has_dot, has_e = False, False
        i, buf = 0, []
        if c == '.':
            buf = ['0']
        while i < TextLen - 1 and (c.isdigit() or c == '.' or c.lower() == 'e'):
            if has_dot and c == '.':
                break
            elif has_e and c.lower() == 'e':
                break
            buf.append(c)
            i += 1
            if c == '.':
                has_dot = True
            elif c.lower() == 'e':
                has_e = True
                c = self.next_char()
                if c == '+' or c == '-':
                    buf.append(c)
                    i += 1
            c = self.next_char()
        self.back_pos()

        word = ''.join(buf)
        token = Token(TokenType.T_NUMLIT)
        token.tok_str = word
        if has_dot or has_e:
            token.num_type = NumType.NUM_FLT
            token.num_val = float(word)
        else:
            token.num_type = NumType.NUM_INT
            token.num_val = int(word)
        return token

    def scan_symbol(self, c: str) -> Token | None:
        two_char_ops = {
            '=': {'=': TokenType.T_EQ},
            '!': {'=': TokenType.T_NE},
            '<': {'=': TokenType.T_LE, '<': TokenType.T_LSHIFT},
            '>': {'=': TokenType.T_GE, '>': TokenType.T_RSHIFT},
            '&': {'&': TokenType.T_LOGAND},
            '|': {'|': TokenType.T_LOGOR}
        }
        if c in two_char_ops:
            next_c = self.next_char()
            if next_c in two_char_ops[c]:
                tok_type = two_char_ops[c][next_c]
                return Token(tok_type)
            self.back_pos()

        single_char_ops = {
            '+': TokenType.T_PLUS,
            '-': TokenType.T_MINUS,
            '*': TokenType.T_STAR,
            '/': TokenType.T_SLASH,
            '%': TokenType.T_MOD,
            '=': TokenType.T_ASSIGN,
            '~': TokenType.T_INVERT,
            '!': TokenType.T_LOGNOT,
            '&': TokenType.T_AMPER,
            '|': TokenType.T_OR,
            '^': TokenType.T_XOR,
            ';': TokenType.T_SEMI,
            '{': TokenType.T_LBRACE,
            '}': TokenType.T_RBRACE,
            '(': TokenType.T_LPAREN,
            ')': TokenType.T_RPAREN,
            ',': TokenType.T_COMMA
        }
        if c in single_char_ops:
            tok_type = single_char_ops[c]
            return Token(tok_type)
        fatal(f"Unknown character {c}")
        return None

    # while self.AtBegin and c == ord('#'):
    #     self.AtBegin = 0
    #     self.scan(Thistoken)
    #     if Thistoken.token != TokenType.T_NUMLIT:
    #         fatal(f"Expecting pre-processor line number, got {Text}")
    #     l = Thistoken.num_val
    #
    #     self.scan(Thistoken)
    #     if Thistoken.token != TokenType.T_STRLIT:
    #         fatal(f"Expecting pre-processor file name, got {Text}")

